frontend/src/pages/QueryData/QueryData.tsx

```diff
@@ -79,7 +81,17 @@ class QueryData extends React.Component<QueryDataProps> {
       }
     }
     if (id) {
-      this.props.actions.fetchOperationData({ id, limit: 10, offset: 0 });
+      localForage.getItem<string>(`${localForageKeys.DATASET_ID}-${id}`).then((data) => {
+        if (data) {
+          this.props.actions.setOperationData(fromJS(JSON.parse(data)), {
+            id,
+            limit: 10,
+            offset: 0,
+          });
+        } else {
+          this.props.actions.fetchOperationData({ id, limit: 10, offset: 0 });
+        }
+      });
```

Comment:
This shouldn't be here ... it should be abstracted under the already existing data fetching/updating code ... so in this case, I only expect local forage code to be in the QueryData saga. i.e. before data is fetched via API, there should be a check for it in the local cache. You should catch the operation's `update` date/time to use it as a check for whether you should fetch new data or go via the cache.